tools/Datadog.Core.Tools/ConcurrencyHelper.cs

```diff
@@ -0,0 +1,152 @@
+
+                                   try
+                                   {
+                                       for (var i = 0; i < level.Iterations; i++)
+                                       {
+                                           try
+                                           {
+                                               level.Action();
+                                           }
+                                           catch (Exception ex)
+                                           {
+                                               level.Exceptions.Add(ex);
+                                           }
+                                       }
+                                   }
+                                   finally
+                                   {
+                                       level.Finished = DateTime.Now;
+                                       level.Status = HelperStatus.Finished;
+                                       Interlocked.Decrement(ref _remainingLevels);
+
+                                       if (_remainingLevels == 0)
+                                       {
+                                           // The run is finished
+                                           _testFinishedEvent.Set();
+                                       }
+                                   }
+                               }));
+
+            foreach (var worker in workers)
+            {
+                registry.Enqueue(worker);
+                worker.Start();
+            }
+
+            // Run everything
+            _initiateLevelsEvent.Set();
+
+            // Wait for everything to finish
+            _testFinishedEvent.Wait();
+
+            Finished = DateTime.Now;
+            Status = HelperStatus.Finished;
+
+            // Pause for the next run if this class is reused
+            _initiateLevelsEvent.Reset();
+
+            await Task.FromResult(0);
+        }
+
+        public Dictionary<string, int> GetExceptionSummary()
+        {
+            return Levels.SelectMany(l => l.Exceptions).GroupBy(ex => ex.Message).ToDictionary(group => group.Key, group => group.Count());
+        }
+
+        public double GetTotalRuntime()
+        {
+            return Levels.Sum(l => l.TotalMilliseconds);
+        }
+
+        public double GetAverageActionRuntime()
+        {
+            return Levels.Average(l => l.AverageCallMilliseconds);
+        }
+
+        public class Level
+        {
+            public string Name { get; set; }
+
+            public Action Action { get; set; }
+
+            public int Iterations { get; set; }
+
+            public ConcurrentBag<Exception> Exceptions { get; set; } = new ConcurrentBag<Exception>();
+
+            public DateTime Started { get; set; }
+
+            public DateTime Finished { get; set; }
+
+            public double TotalMilliseconds => (Finished - Started).TotalMilliseconds;
```

Comment:
We should use [`Stopwatch`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch) instead of `DateTime`. It is specifically designed for uses like this (measuring elapsed time in benchmarks) and uses high-resolution hardware if available.

https://stackoverflow.com/a/28648/24231

Reply:
This is still subtracting `DateTime` instances. Do we still use `ConcurrencyHelper` anywhere?

Reply:
No, but I want to use it, without the timers in integration tests in future changes. I can remove this DateTime stuff for now.